### PR TITLE
Added conditional for initial_lifecycle_hook parameter

### PR DESCRIPTION
**Pull request by robh007**
I proposed adding the terraform resource `aws_autoscaling_lifecycle_hook` into this module. @antonbabenko suggested creating a submodule to allow this, this is still outstanding. However this PR allows you to create an `aws_autoscaling_group` with the parameter `initial_lifecycle_hook`. The reason behind this is that if you use the `aws_autoscaling_lifecycle_hook` resource when issuing a terraform destroy the lifecycle hook is removed from the autoscaling group before any lifecycle transition occurs. This is an issue when creating events based on instance termination.

In my usecase we are using the servicenaming API to register instances into Route53, the lifecycle hook is used to remove Route53 records if / when instances are terminated. So issuing a terraform destroy, terraform removes the instance from the ASG prior to any lifecycle transition. If you use the `initial_lifecycle_hook` parameter within the autoscaling group resource, the ASG transitions as expected and waits for the terminate lifecycle action to complete before destroying the ASG.

I had to create a separate resource because the `initial_lifecycle_hook` as some required parameters. I tried to use a similar pattern to how the tags work, but due to the current hit & miss with maps and lists within the HCL language it didn't work.

I've updated the README with the new variables and also created an additional example.

### CODE REVIEW

**Comment by thenom** [comment]
+1

**Comment by robh007** [comment]
@antonbabenko any possibility of this being reviewed?

**Comment by antonbabenko** [comment]
Hi @robh007 ! Thanks for your contribution. I won't be able to review it properly and release during next 30 days because of vacation in the middle of nowhere. If you need this, please use your fork in the meanwhile.

File: main.tf [hunk 1]
```diff
@@ -66,6 +66,57 @@ resource "aws_autoscaling_group" "this" {
...
+  enabled_metrics           = ["${var.enabled_metrics}"]
+  metrics_granularity       = "${var.metrics_granularity}"
+  wait_for_capacity_timeout = "${var.wait_for_capacity_timeout}"
+  protect_from_scale_in     = "${var.protect_from_scale_in}"
+
+  initial_lifecycle_hook {
+    name                    = "${var.initial_lifecycle_hook_name}"
+    lifecycle_transition    = "${var.initial_lifecycle_hook_transition}"
+    notification_metadata   = "${var.initial_lifecycle_hook_notification_metadata}"
+    heartbeat_timeout       = "${var.initial_lifecycle_hook_heartbeat_time}"
```
Comment (antonbabenko) [inline main.tf]: Rename var to `initial_lifecycle_hook_heartbeat_timeout`

File: main.tf [hunk 2]
```diff
@@ -66,6 +66,57 @@ resource "aws_autoscaling_group" "this" {
...
+  suspended_processes       = "${var.suspended_processes}"
+  placement_group           = "${var.placement_group}"
+  enabled_metrics           = ["${var.enabled_metrics}"]
+  metrics_granularity       = "${var.metrics_granularity}"
+  wait_for_capacity_timeout = "${var.wait_for_capacity_timeout}"
+  protect_from_scale_in     = "${var.protect_from_scale_in}"
+
+  initial_lifecycle_hook {
+    name                    = "${var.initial_lifecycle_hook_name}"
+    lifecycle_transition    = "${var.initial_lifecycle_hook_transition}"
```
Comment (antonbabenko) [inline main.tf]: Rename to `initial_lifecycle_hook_lifecycle_transition`

File: variables.tf [hunk 3]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
   default     = true
 }
 
+variable "create_asg_with_initial_lifecycle_hook" {
+  description = "Create an ASG with intial"
```
Comment (antonbabenko) [inline variables.tf]: "with initial lifecycle hook"

File: main.tf [hunk 4]
```diff
@@ -66,6 +66,57 @@ resource "aws_autoscaling_group" "this" {
   }
 }
 
+################################################
+# Autoscaling group with initial lifecycle hook
+################################################
+resource "aws_autoscaling_group" "with_initial_lifecycle_hook" {
+  count = "${var.create_asg_with_initial_lifecycle_hook}"
+
+  name_prefix          = "${join("-", compact(list(coalesce(var.asg_name, var.name), var.recreate_asg_when_lc_changes ? element(concat(random_pet.asg_name.*.id, list("")), 0) : "")))}-"
```
Comment (antonbabenko) [inline main.tf]: Please rebase to get recent changes from the master branch

File: variables.tf [hunk 5]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
...
+  default     = false
+}
+
+variable "initial_lifecycle_hook_name" {
+  description = "Initial lifecycle hook name required parameter"
```
Comment (antonbabenko) [inline variables.tf]: "The name of initial lifecycle hook"

File: variables.tf [hunk 6]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
...
+  default     = ""
+}
+
+variable "initial_lifecycle_hook_transition" {
+  description = "Initial Lifecycle hook transtion type"
```
Comment (antonbabenko) [inline variables.tf]: "The instance state to which you want to attach the initial lifecycle hook"

File: variables.tf [hunk 7]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
...
+  default     = ""
+}
+
+variable "initial_lifecycle_hook_default_result" {
+  description = "Defines the default action the autoscaler takes when the lifecyle hook timesout"
```
Comment (antonbabenko) [inline variables.tf]: "Defines the action the Auto Scaling group should take when the lifecycle hook timeout elapses or if an unexpected failure occurs. The value for this parameter can be either CONTINUE or ABANDON. The default value for this parameter is ABANDON."

File: variables.tf [hunk 8]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
...
+  default     = "CONTINUE"
+}
+
+variable "initial_lifecycle_hook_notification_metadata" {
```
Comment (antonbabenko) [inline variables.tf]: "Contains additional information that you want to include any time Auto Scaling sends a message to the notification target"

File: variables.tf [hunk 9]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
...
+  description = "Metadata for the lifecycle notification event"
+  default     = ""
+}
+
+variable "initial_lifecycle_hook_heartbeat_time" {
+  description = "Lifecyle action timer"
```
Comment (antonbabenko) [inline variables.tf]: "Defines the amount of time, in seconds, that can elapse before the lifecycle hook times out. When the lifecycle hook times out, Auto Scaling performs the action defined in the DefaultResult parameter"

File: variables.tf [hunk 10]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
...
+  default     = "60"
+}
+
+variable "initial_lifecycle_hook_notification_target_arn" {
+  description = "SNS target arn to allow notifications from autoscaling group"
```
Comment (antonbabenko) [inline variables.tf]: "The ARN of the notification target that Auto Scaling will use to notify you when an instance is in the transition state for the lifecycle hook. This ARN target can be either an SQS queue or an SNS topic."

File: variables.tf [hunk 11]
```diff
@@ -8,6 +8,46 @@ variable "create_asg" {
...
+  default     = ""
+}
+
+variable "initial_lifecycle_hook_role_arn" {
+  description = "AWS iam role to grant access to notification target"
```
Comment (antonbabenko) [inline variables.tf]: "The ARN of the IAM role that allows the Auto Scaling group to publish to the specified notification target."

File: examples/asg_inital_lifecycle_hook/README.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# Auto Scaling Group without ELB example
```
Comment (antonbabenko) [inline examples/asg_inital_lifecycle_hook/README.md]: "Auto Scaling Group with initial lifecycle hook"

Comment (antonbabenko) [reply examples/asg_inital_lifecycle_hook/README.md]: Also, add a link to this example in the main README.md file.

Comment (robh007) [reply examples/asg_inital_lifecycle_hook/README.md]: @antonbabenko I'm away at the moment. I will try and get these issues sorted.

**Comment by robh007** [comment]
I think that's the changes you requested. Apologies for delay and thanks to @thenom.

**Comment by antonbabenko** [comment]
@robh007 I will review it on Friday.

Sorry for the delay! Life and work takes more time than before :)

**Comment by antonbabenko** [comment]
Thanks everyone for the patience!

[v2.9.0](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/releases/v2.9.0) has been released as part of #52 (which is based on this PR).

**Pull request closed without merging**